crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -485,12 +485,9 @@ impl WritebackCache {
         //   the tx finalizer, plus checkpoint executor, consensus, and RPCs from fullnodes.
         let mut entry = self.dirty.objects.entry(*object_id).or_default();
 
-        self.cached.object_by_id_cache.insert(
-            *object_id,
-            Arc::new(Mutex::new(LatestObjectCacheEntry::Object(
-                version,
-                object.clone(),
-            ))),
+        self.insert_cache_mononotic(
+            object_id,
+            LatestObjectCacheEntry::Object(version, object.clone()),
```

Comment:
By the way, if you're wondering "how could I have been so stupid as to think the old code could ever work", my thinking depended on the fact that the write here is guaranteed to have the maximum version yet observed anywhere in the system, hence it would always win in the `is_newer_than` check.

Reply:
Would it be worthwhile to encapsulate the object by id cache to a different file with a new type, so that you could hide `insert` entirely? Just to make sure we don't ever accidentally call `insert`.

Reply:
Not a bad idea. I may just use a module within the current file though.